### PR TITLE
feat(paymaster-proxy): add runtime migrations

### PR DESCRIPTION
**Pull request by jakim929**
closes https://github.com/ethereum-optimism/ecopod/issues/950

- updates envvars parsing to use znv
- runs migrations at runtime
- updates readiness check to depend on whether migration has run or not

### CODE REVIEW

**Comment by jakim929** [comment]
> [!WARNING]
> <b>This pull request is not mergeable via GitHub because a downstack PR is open. Once all requirements are satisfied, merge this PR as a stack <a href="https://app.graphite.dev/github/pr/ethereum-optimism/ecosystem/288?utm_source=stack-comment-downstack-mergeability-warning" >on Graphite</a>.</b>
> <a href="https://graphite.dev/docs/merge-pull-requests">Learn more</a>

* **#288** <a href="https://app.graphite.dev/github/pr/ethereum-optimism/ecosystem/288?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈
* **#278** <a href="https://app.graphite.dev/github/pr/ethereum-optimism/ecosystem/278?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#268** <a href="https://app.graphite.dev/github/pr/ethereum-optimism/ecosystem/268?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#260** <a href="https://app.graphite.dev/github/pr/ethereum-optimism/ecosystem/260?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#256** <a href="https://app.graphite.dev/github/pr/ethereum-optimism/ecosystem/256?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#255** <a href="https://app.graphite.dev/github/pr/ethereum-optimism/ecosystem/255?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#230** <a href="https://app.graphite.dev/github/pr/ethereum-optimism/ecosystem/230?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* `main`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @jakim929 and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment">Graphite <b>Graphite</b></a>

File: apps/paymaster-proxy/package.json [hunk 1]
```diff
@@ -10,16 +10,16 @@
     "start": "node dist/runProxyService.js",
     "lint": "eslint \"**/*.{ts,tsx}\" && pnpm prettier --check \"**/*.{ts,tsx}\"",
     "lint:fix": "eslint \"**/*.{ts,tsx}\" --fix --quiet && pnpm prettier \"**/*.{ts,tsx}\" --write --loglevel=warn",
-    "generate-example-env": "ts-node scripts/generateExampleEnv.ts",
     "loadtest": "pnpm dlx artillery run loadtest/basic-loadtest.yml",
-    "docker:services:up": "docker compose up -d redis postgres",
-    "docker:up": "docker compose up -d redis postgres paymaster-proxy-app",
+    "docker:services:up": "nx run api-key-service:docker:up && docker compose up -d redis postgres",
+    "docker:up": "nx run api-key-service:docker:up && docker compose up -d redis postgres paymaster-proxy-app",
```
Comment (jakim929) [inline apps/paymaster-proxy/package.json:15]: not ideal but will keep doing this until we unify the monorepo into a single docker compose

File: Dockerfile [hunk 2]
```diff
@@ -32,8 +32,8 @@ COPY . ./
 # install monorepo dependencies, from the virtual store
 RUN --mount=type=cache,id=pnpm,target=/pnpm/store pnpm install --frozen-lockfile --prefer-offline
 
-# build all packages/apps
-RUN pnpm nx run-many --target=build
+# build all apps we want to deploy
+RUN pnpm nx run-many --target=build --projects=paymaster-proxy,dapp-console-api,api-key-service
```
Comment (jakim929) [inline Dockerfile:36]: the dapp-console frontend nextjs build was taking up most of the time, so excluding it from our docker builds

File: apps/paymaster-proxy/src/helpers/BackendReadinessState.ts [hunk 3]
```diff
@@ -0,0 +1,11 @@
+export class BackendReadinessState {
```
Comment (jakim929) [inline apps/paymaster-proxy/src/helpers/BackendReadinessState.ts:1]: used to expose readiness from the ProxyService to the apiServer

File: apps/paymaster-proxy/package.json [hunk 4]
```diff
@@ -10,16 +10,16 @@
...
     "docker:replicas:up": "docker compose up -d redis postgres paymaster-proxy-app-replica reverse-proxy",
     "docker:down": "docker compose down",
     "drizzle:studio": "drizzle-kit studio",
     "migrations:generate": "drizzle-kit generate:pg",
     "migrations:execute": "ts-node ./scripts/db-migrate",
-    "migrations:dev": "pnpm migrations:generate && pnpm migrations:execute"
+    "migrations:dev": "pnpm migrations:generate && pnpm migrations:execute",
+    "env:example:generate": "NODE_ENV=test ts-node scripts/generate-example-env.ts"
```
Comment (tremarkley) [inline apps/paymaster-proxy/package.json:22]: oooo I like this

File: apps/paymaster-proxy/scripts/generate-example-env.ts
Comment (tremarkley) [inline apps/paymaster-proxy/scripts/generate-example-env.ts:1]: might be worth adding this to the nx generator we have for creating a trpc api service!

**Review by tremarkley — APPROVED** [review]

File: apps/paymaster-proxy/scripts/generate-example-env.ts
Comment (jakim929) [reply apps/paymaster-proxy/scripts/generate-example-env.ts:1]: sounds good, will add a PR for this

**Pull request merged**